### PR TITLE
Refactor list renumbering logic

### LINKED ISSUE #56: Refactor list renumbering logic to use stack and hash map for improved readability and maintainability
The current implementation of the list renumbering logic is complex, involving manual stack management, nested `is_some_and` checks, and two loops for handling indentation and counters. This makes the code harder to read and maintain.

**Suggested Refactor:**
- Extract the numbered-line parsing logic into a small helper function for clarity.
- Use a `Vec<usize>` as an indentation stack and a `HashMap<indent, next>` for tracking counters at each indentation level.
- Replace the two loops for dropping deeper indents with a single concise `while` loop and `HashMap::remove`.
- Use `entry(...).or_insert(1)` to seed new lists, simplifying counter initialization.

This approach will:
- Remove the need for a manual `(usize, usize)` stack in favor of a single `Vec<usize>` and `HashMap`.
- Collapse the two `while` loops into one.
- Improve readability by pulling out regex logic into a helper.
- Preserve all existing functionality, including code fence toggling, passthrough for code blocks, and cross-block counting.

**Example Implementation:**
```rust
use std::collections::HashMap;
use lazy_static::lazy_static;
use regex::Regex;

lazy_static! {
    static ref FENCE_RE: Regex     = Regex::new(r"^(```|~~~).*").unwrap();
    static ref NUMBERED_RE: Regex  = Regex::new(r"^(\s*)([1-9][0-9]*)\.(\s+)(.*)").unwrap();
}

fn parse_numbered(line: &str) -> Option<(usize, &str, &str)> {
    let caps = NUMBERED_RE.captures(line)?;
    let indent = caps.get(1)?.as_str().len();
    let sep    = caps.get(3)?.as_str();
    let rest   = caps.get(4)?.as_str();
    Some((indent, sep, rest))
}

pub fn renumber_lists(lines: &[String]) -> Vec<String> {
    let mut out     = Vec::with_capacity(lines.len());
    let mut stack   = Vec::<usize>::new();
    let mut counters = HashMap::<usize, usize>::new();
    let mut in_code = false;

    for line in lines {
        if FENCE_RE.is_match(line) {
            in_code = !in_code;
            out.push(line.clone());
            continue;
        }
        if in_code {
            out.push(line.clone());
            continue;
        }
        if let Some((indent, sep, rest)) = parse_numbered(line) {
            // drop all deeper-indent lists
            while stack.last().map_or(false, |&d| d > indent) {
                counters.remove(&stack.pop().unwrap());
            }
            // start new list at this indent if needed
            if stack.last().map_or(true, |&d| d < indent) {
                stack.push(indent);
                counters.insert(indent, 1);
            }
            // grab & bump our counter
            let num = counters.get_mut(&indent).unwrap();
            let current = *num;
            *num += 1;

            out.push(format!(
                "{}{}.{}{}",
                " ".repeat(indent), current, sep, rest
            ));
        } else {
            out.push(line.clone());
        }
    }
    out
}
```

**Action Items:**
- Refactor the list renumbering logic as described above.
- Ensure all existing tests pass and add new tests if necessary to cover edge cases.
- Review for readability and maintainability.

---

I created this issue for @leynos from https://github.com/leynos/mdtablefix/pull/54#discussion_r2203383411.

<details>
<summary>Tips and commands</summary>

#### Getting Help

- [Contact our support team](mailto:support@sourcery.ai) for questions or feedback.
- Visit our [documentation](https://docs.sourcery.ai) for detailed guides and information.
- Keep in touch with the Sourcery team by following us on [X/Twitter](https://x.com/SourceryAI), [LinkedIn](https://www.linkedin.com/company/sourcery-ai/) or [GitHub](https://github.com/sourcery-ai).

</details>

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- simplify ordered list renumbering using indentation stack and HashMap counters
- drop unused counter helpers and related tests

## Testing
- `make fmt`
- `make lint`
- `make test`

closes #56

------
https://chatgpt.com/codex/tasks/task_e_68c2b61debc4832281f9094856ad6c86

## Summary by Sourcery

Refactor ordered list renumbering to use an indentation stack and HashMap for counters, dropping obsolete helpers and related tests

Enhancements:
- Replace vector-based counters with an indentation stack and HashMap to track numbering levels
- Update parse_numbered to calculate numeric indentation and return both raw and computed indent
- Simplify paragraph restart and deeper-level removal logic using stack operations and HashMap removal
- Clear stack and counters upon encountering headings or thematic breaks

Tests:
- Remove deprecated pop_counters_upto tests and adjust imports for renumber_lists

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai pause

File: src/lists.rs [hunk 1]
```diff
@@ -88,57 +75,55 @@
...
+                if let Some(d) = stack.pop() {
+                    counters.remove(&d);
                 }
-                _ => {
-                    counters.push((indent, 1));
-                    1
-                }
-            };
+            }
+            if stack.last().is_none_or(|&d| d < indent) {
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/lists.rs]: Has this now been resolved?

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

- Extract the repeated “while stack.last() > indent { pop and remove counters }” logic into a single helper function to reduce duplication and improve readability.
- Rename the generic `stack` variable to something like `indent_stack` so its role in tracking list indent levels is clearer.
- Since `parse_numbered` already computes both `indent_str` and its length, consider returning both directly there to avoid calling `indent_len` again immediately afterward.

File: src/lists.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (leynos) [reply src/lists.rs]: Has this now been resolved?
